`Integrated code lifecycle`: Allow adjustment of # concurrent build jobs during runtime

ekayandan opened the pull request:
…h new DTO and REST endpoints for concurrent build size adjustment

<!-- Thanks for contributing to Artemis! Before you submit your pull request, please make sure to check all tasks by putting an x in the [ ] (don't: [x ], [ x], do: [x]). Remove not applicable tasks and do not leave them unchecked -->
<!-- If your pull request is not ready for review yet, create a draft pull request! -->

### Checklist
#### General
<!-- Remove tasks that are not applicable for your PR. Please only put the PR into ready for review, if all relevant tasks are checked! -->
<!-- You only need to choose one of the first two check items: Generally, test on the test servers. -->
<!-- If it's only a small change, testing it locally is acceptable, and you may remove the first checkmark. If you are unsure, please test on the test servers. -->
- [ ] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [ ] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [ ] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [ ] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/development-process.html#naming-conventions-for-github-pull-requests).


#### Server
- [ ] **Important**: I implemented the changes with a [very good performance](https://docs.artemis.cit.tum.de/dev/guidelines/performance/) and prevented too many (unnecessary) and too complex database calls.
- [ ] I **strictly** followed the principle of **data economy** for all database calls.
- [ ] I **strictly** followed the [server coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [ ] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [ ] I added pre-authorization annotations according to the [guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/#rest-endpoint-best-practices-for-authorization) and checked the course groups for all new REST Calls (security).
- [ ] I documented the Java code using JavaDoc style.


#### Client
- [ ] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data (e.g. using paging).
- [ ] I **strictly** followed the principle of **data economy** for all client-server REST calls.
- [ ] I **strictly** followed the [client coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [ ] Following the [theming guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [ ] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [ ] I added `authorities` to all new routes and checked the course groups for displaying navigation elements (links, buttons).
- [ ] I documented the TypeScript code using JSDoc style.
- [ ] I added multiple screenshots/screencasts of my UI changes.
- [ ] I translated all newly inserted strings into English and German.


#### Changes affecting Programming Exercises
- [ ] **High priority**: I tested **all** changes and their related features with **all** corresponding user types on a test server configured with the **integrated lifecycle setup** (LocalVC and LocalCI).
- [ ] I tested **all** changes and their related features with **all** corresponding user types on a test server configured with **LocalVC** and **Jenkins**.


### Motivation and Context
<!-- Why is this change required? What problem does it solve? -->
<!-- If it fixes an open issue, please link to the issue here. -->


### Description
<!-- Describe your changes in detail -->


### Steps for Testing
<!-- Please describe in detail how reviewers can test your changes. Make sure to take all related features and views into account! Below is an example that you can refine. -->
Prerequisites:
- 1 Instructor
- 2 Students
- 1 Programming Exercise with Complaints enabled

1. Log in to Artemis
2. Navigate to Course Administration
3. ...

#### Exam Mode Testing
<!-- If this PR changes some components that are also used in the exam mode, the PR needs additional testing that the exam mode is still working as expected. -->
<!-- If the testing steps above already describe the exam mode or the exam mode cannot be affected by this PR in any way, you can leave this out. -->

Prerequisites:
- 1 Instructor
- 2 Students
- 1 Exam with a Programming Exercise

1. Log in to Artemis
2. Participate in the exam as a student
3. Make sure that the UI of the programming exercise in the exam mode stays unchanged. You can use the [exam mode documentation](https://docs.artemis.cit.tum.de/user/exam_mode/) as reference.
4. ...

### Testserver States
You can manage test servers using [Helios](https://helios.aet.cit.tum.de/). Check environment statuses in the [environment list](https://helios.aet.cit.tum.de/repo/69562331/environment/list). To deploy to a test server, go to the [CI/CD](https://helios.aet.cit.tum.de/repo/69562331/ci-cd) page, find your PR or branch, and trigger the deployment.

### Review Progress
<!-- Each PR should be reviewed by at least two other developers. The code, the functionality (= manual test) and the exam mode need to be reviewed. -->
<!-- The reviewer or author check the following boxes depending on what was reviewed or tested. All boxes should be checked before merge. -->
<!-- You can add additional checkboxes if it makes sense to only review parts of the code or functionality. -->
<!-- When changes are pushed, uncheck the affected boxes. (Not all changes require full re-reviews.) -->
<!-- All PRs that might affect the exam mode (e.g. change a client component that is also used in the exam mode) need an additional verification that the exam mode still works. -->

#### Performance Review
<!-- See [Large Course Setup](https://github.com/ls1intum/Artemis/tree/develop/supporting_scripts/course-scripts/quick-course-setup) for the automation script that handles large course setup. -->
- [ ] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance even for very large courses with more than 2000 students.
- [ ] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance even for very large courses with more than 2000 students.
#### Code Review
- [ ] Code Review 1
- [ ] Code Review 2
#### Manual Tests
- [ ] Test 1
- [ ] Test 2
#### Exam Mode Test
- [ ] Test 1
- [ ] Test 2
#### Performance Tests
- [ ] Test 1
- [ ] Test 2

### Test Coverage
<!-- Please add the test coverages for all changed files modified in this PR here. You can use `supporting_script/code-coverage/generate_code_cov_table/generate_code_cov_table.py` to automatically generate the coverage table from the corresponding artefacts of your branch (follow the ReadMe for setup details). -->
<!-- Alternatively you can execute the tests locally (see build.gradle and package.json) or look into the corresponding artefacts. -->
<!-- The line coverage must be above 90% for changes files, and you must use extensive and useful assertions for server tests and expect statements for client tests. -->
<!-- Note: Confirm in the last column that you have implemented extensive assertions for server tests and expect statements for client tests. -->
<!--       Remove rows with only trivial changes from the table. -->
<!--
| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|-----------------------------:|
| ExerciseService.java | 85% | ✅                           |
| programming-exercise.component.ts | 95% | ✅              |
-->

### Screenshots
<!-- Add screenshots to demonstrate the changes in the UI. Remove the section if you did not change the UI. -->
<!-- Create a GIF file from a screen recording in a docker container https://toub.es/2017/09/11/high-quality-gif-with-ffmpeg-and-docker/ -->

konrad2002 reviewed (approved):
tested on TS3, works as expected

konrad2002 reviewed (changes requested):
nvm, when the agents have build jobs running, then pausing them, changing the "Max # of concurrent build jobs" and then resuming them leads to the build agents still accepting the previous amount of builds. Even after waiting for all of them to be idle and re-commiting, see picture

<img width="2862" height="1102" alt="grafik" src="https://github.com/user-attachments/assets/636c1141-ca16-4d2e-9f15-cee97b7714b3" />

ekayandan commented:
> tested on TS3, everything works when not pausing the agents; however when the agents have build jobs running, then pausing them, changing the "Max # of concurrent build jobs" and then resuming them leads to the build agents still accepting the previous amount of builds. Even after waiting for all of them to be idle and re-commiting, see picture
> 
> <img alt="grafik" width="2000" height="1102" src="https://private-user-images.githubusercontent.com/44098768/468424593-636c1141-ca16-4d2e-9f15-cee97b7714b3.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTMxMzQwMTYsIm5iZiI6MTc1MzEzMzcxNiwicGF0aCI6Ii80NDA5ODc2OC80Njg0MjQ1OTMtNjM2YzExNDEtY2ExNi00ZDJlLTlmMTUtY2VlOTdiNzcxNGIzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA3MjElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNzIxVDIxMzUxNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWJhZjNiYjQ3YmZlYjIwN2U5ZTQ4MDEwNTRlZjQ5NWM2MzUxYzg5NmIxNGMxYzgwN2RkMzk2NGQxN2ZjODhjNjkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.WtkJB0woKXsCPnJ_R9b5kMO_lTTgC8wPe2Jf1ifePoE">

I have disabled concurrency adjustment while agent is paused. This issue should not occur anymore 😅

jfr2102 commented on src/main/java/de/tum/cit/aet/artemis/core/web/admin/AdminBuildJobQueueResource.java:
```diff
@@ -280,6 +281,28 @@ public ResponseEntity<Void> resumeAllBuildAgents() {
+            throw new BadRequestAlertException("Agent name cannot be null or empty", "agentName", "agentNameInvalid");
+        }
+        if (newSize <= 0) {
+            throw new BadRequestAlertException("Concurrent build size must be positive", "newSize", "invalidConcurrentSize");
```
```suggestion
            throw new BadRequestAlertException("Concurrent build size must be at least 1", "newSize", "invalidConcurrentSize");
```

jfr2102 commented on src/main/java/de/tum/cit/aet/artemis/core/web/admin/AdminBuildJobQueueResource.java:
```diff
@@ -280,6 +281,28 @@ public ResponseEntity<Void> resumeAllBuildAgents() {
+            throw new BadRequestAlertException("Concurrent build size must be positive", "newSize", "invalidConcurrentSize");
+        }
+
+        localCIBuildJobQueueService.adjustBuildAgentCapacity(agentName, newSize);
```
we should throw a BadRequest for newSize large than configurable max (see my other comments) too I think

jfr2102 replied:
Either here, or move it in localCIBuildJobQueueService.adjustBuildAgentCapacity()

ekayandan replied:
moved to service 👍

jfr2102 commented on src/main/java/de/tum/cit/aet/artemis/buildagent/BuildAgentConfiguration.java:
```diff
@@ -149,21 +183,17 @@ else if (bytes < 1024 * 1024 * 1024) {
      *
      * @return The executor service.
      */
     private ThreadPoolExecutor createBuildExecutor() {
```
we currently have `artemis.continuous-integration.concurrent-build-size` and `artemis.continuous-integration.specify-concurrent-builds` configuration values. 

I would suggest to replace `artemis.continuous-integration.concurrent-build-size` by 2 config values, one for the default, one for maximum.
This maximum should then be enforced on server side + on client side i think. ( see other comments)

If specify-concurrent-builds is false, we could opt for a default for the maximum like the `numberOfCores` which you currently set as min in the client side. But it should still be checked server side too.

ekayandan replied:
That's actually a good solution. I was planning to implement a way of preventing agents to pull jobs after a certain memory threshold, so I added this as a temporary solution. But we definitely should allow configuration of max value as well. I'm on it!

jfr2102 commented on src/main/webapp/app/buildagent/build-agent-summary/build-agent-summary.component.html:
```diff
@@ -82,15 +82,23 @@ <h3 id="build-agents-heading" jhiTranslate="artemisApp.buildAgents.summary"></h3
+                                <jhi-number-input
+                                    [value]="concurrencyMap[row.buildAgent?.name] || value"
+                                    [minValue]="1"
+                                    [maxValue]="row.numberOfCores || 8"
```
Id set the configurable max here as max, and on the server side we can set that value to the numberOfCores if `specify-concurrent-builds` is false. Id also change the type to not be possible undefined, this should not happen right?

jfr2102 commented on src/main/java/de/tum/cit/aet/artemis/buildagent/dto/BuildAgentInformation.java:
```diff
@@ -16,7 +16,7 @@
 public record BuildAgentInformation(@NotNull BuildAgentDTO buildAgent, int maxNumberOfConcurrentBuildJobs, int numberOfCurrentBuildJobs,
         @NotNull List<BuildJobQueueItem> runningBuildJobs, @Nullable BuildAgentStatus status, String publicSshKey, @Nullable BuildAgentDetailsDTO buildAgentDetails,
-        int pauseAfterConsecutiveBuildFailures) implements Serializable {
+        int pauseAfterConsecutiveBuildFailures, int numberOfCores) implements Serializable {
```
maybe instead of numberOfCores `maxConfigurableConcurrentBuildJobs` or so?

jfr2102 commented on src/main/webapp/app/buildagent/shared/entities/build-agent-information.model.ts:
```diff
@@ -19,6 +19,7 @@ export class BuildAgentInformation implements BaseEntity {
     public recentBuildJobs?: BuildJob[];
     public buildAgentDetails?: BuildAgentDetails;
     public pauseAfterConsecutiveBuildFailures?: number;
+    public numberOfCores?: number;
```
maybe this, and other primitives can typed without "?" -> no undefined? (Or will 0 ints be currently be left out in json serialization? for the BuildAgentDetailsDTO?)

jfr2102 reviewed (changes requested):
Left some a few comments regarding configuration and setting the maximum number for the threadpool.

I think another useful addition could be E2E tests for this feature, but it could be added later too imo.

Jan-Thurner reviewed (approved):
Tested on TS3, works as expected

jfr2102 commented on src/main/java/de/tum/cit/aet/artemis/buildagent/service/SharedQueueProcessingService.java:
```diff
@@ -167,6 +168,29 @@ public void init() {
+
+        distributedDataAccessService.getAdjustBuildAgentCapacityTopic().addMessageListener(message -> {
+            BuildAgentCapacityAdjustmentDTO adjustment = message.getMessageObject();
+            if (adjustment.getBuildAgentName() == null || buildAgentShortName.equals(adjustment.getBuildAgentName())) {
```
the agent name cannot be null here I think 
Also I would suggest to annotate in BuildAgentCapacityAdjustmentDTO accordingly for non primitives

ekayandan replied:
Oh yes, I initially had null as a broadcast flag. but it can't be null anymore. removed.

jfr2102 commented on src/main/java/de/tum/cit/aet/artemis/buildagent/dto/BuildAgentCapacityAdjustmentDTO.java:
```diff
@@ -0,0 +1,24 @@
+
+    private final int newCapacity;
+
+    public BuildAgentCapacityAdjustmentDTO(String buildAgentName, int newCapacity) {
```
```suggestion
    public BuildAgentCapacityAdjustmentDTO(@NotNull String buildAgentName, int newCapacity) {
```

ekayandan replied:
Done

jfr2102 commented on src/main/java/de/tum/cit/aet/artemis/buildagent/dto/BuildAgentCapacityAdjustmentDTO.java:
```diff
@@ -0,0 +1,24 @@
+package de.tum.cit.aet.artemis.buildagent.dto;
+
+/** DTO for build agent capacity adjustment messages. */
+public class BuildAgentCapacityAdjustmentDTO {
```
similar to other DTOs that are shared via Hazelcast, maybe add a comment mentioning that changes to it are potentially breaking changes

ekayandan replied:
done

jfr2102 reviewed (commented):
Looks good overall, thanks for adding the configuration for maximum and default parallel jobs! 

I made 2 more small comments inline
And the PR should mention somewhere that changes to the DTOs that are shared via hazelcast are made, which is a breaking change in theory. 

(If one would run different versions of build agents, or agent vs. core nodes etc.)

ekayandan commented on src/test/java/de/tum/cit/aet/artemis/buildagent/service/BuildAgentIntegrationTest.java:
```diff
@@ -84,6 +89,14 @@ void clearQueues() {
+        // Initialize build agent information in Hazelcast map
+        String memberAddress = hazelcastInstance.getCluster().getLocalMember().getAddress().toString();
+        BuildAgentDTO buildAgentDTO = new BuildAgentDTO(buildAgentShortName, memberAddress, buildAgentDisplayName);
+        BuildAgentInformation initialBuildAgent = new BuildAgentInformation(buildAgentDTO, 2, 0, new ArrayList<>(), BuildAgentInformation.BuildAgentStatus.IDLE, null, null,
+                pauseAfterConsecutiveFailures, 2);
```
Method usage is correct

timn21 reviewed (commented):
If input is beyond the allowed range, we set it to closest edge(min or max). But probably it should be communicated more clearly to user or invalid should be ignored. I will consider options and update here 
Tested on TS3. High inputs are denied, and you can't change the input when agents are paused, this should cover the most obvious edge cases. Everything as expected 👍 
Only one small note: When the input is denied, you still get a success popup, which is a bit counterintuitive.
<img width="1695" height="535" alt="image" src="https://github.com/user-attachments/assets/75370479-baf0-4268-9b6d-0c66fdabc67b" />

ekayandan commented on src/main/java/de/tum/cit/aet/artemis/buildagent/dto/BuildAgentCapacityAdjustmentDTO.java:
```diff
@@ -0,0 +1,25 @@
+    private final int newCapacity;
+
+    public BuildAgentCapacityAdjustmentDTO(String buildAgentName, int newCapacity) {
+        if (newCapacity <= 0)
+            throw new IllegalArgumentException("New capacity must be greater than 0");
```
Outdated. we have @notnull annotation for name now

ekayandan commented on src/main/java/de/tum/cit/aet/artemis/buildagent/dto/BuildAgentCapacityAdjustmentDTO.java:
```diff
@@ -0,0 +1,27 @@
+
+// NOTE: this data structure is used in shared code between core and build agent nodes. Changing it requires that the shared data structures in Hazelcast (or potentially Redis)
+// in the future are migrated or cleared. Changes should be communicated in release notes as potentially breaking changes./
+public class BuildAgentCapacityAdjustmentDTO {
```
converting

ekayandan commented on src/main/java/de/tum/cit/aet/artemis/programming/service/localci/SharedQueueManagementService.java:
```diff
@@ -111,6 +114,42 @@ public void resumeAllBuildAgents() {
+        }
+        catch (Exception e) {
+            log.error("Error adjusting build agent capacity for agent {} to {}", agentName, newCapacity, e);
+            throw new ResponseStatusException(HttpStatus.INTERNAL_SERVER_ERROR, "Failed to adjust build agent capacity");
```
done

ekayandan commented on src/main/webapp/app/buildagent/shared/number-input/number-input.component.ts:
```diff
@@ -0,0 +1,62 @@
+    selector: 'jhi-number-input',
+    templateUrl: './number-input.component.html',
+    styleUrl: './number-input.component.scss',
+    standalone: true,
```
removed

ekayandan commented on src/main/webapp/app/buildagent/shared/number-input/number-input.component.ts:
```diff
@@ -0,0 +1,62 @@
+    templateUrl: './number-input.component.html',
+    styleUrl: './number-input.component.scss',
+    standalone: true,
+    imports: [FontAwesomeModule, CommonModule],
```
done

krusche reviewed (changes requested):
Make clear in the user interface, that the changed values are not taken into account when the whole Hazelcast cluster restarts, e.g. when deploying a new artemis version or in case of failure

tobias-lippert reviewed (approved):
code

florian-glombik reviewed (changes requested):
Tested on ts3

Can we please remove the success toast on every update? For me as a user it is quite annoying, I do already get the immediate feedback in the UI that the value was saved, I do not need a toast in addition to it (which also blocks further adjustments) - if I do not receive an error toast I assume the value to be saved

I also managed to crash the server 
```

2025-08-07T18:40:03.119Z  INFO 1 --- [Artemis] [at-handler-2920] d.t.c.a.a.p.service.GitService           : Pull /opt/artemis/data/repos/GE25BASTESTCOURSETEST/ge25bastestcoursetest-ge25bas |  
-- | --
  | 2025-08-07T18:40:03.196Z ERROR 1 --- [Artemis] [at-handler-2920] d.t.c.a.a.p.service.AbstractGitService   : Cannot use ssh properly because of mismatch of Jgit transport object: org.eclipse.jgit.transport.TransportLocal@4674f753 |  
  | 2025-08-07T18:40:12.973Z ERROR 1 --- [Artemis] [at-handler-2920] d.t.c.a.a.p.service.AbstractGitService   : Cannot use ssh properly because of mismatch of Jgit transport object: org.eclipse.jgit.transport.TransportLocal@688dfc86 |  
  | 2025-08-07T18:40:20.588Z  WARN 1 --- [Artemis] [coveryClient-%d] c.netflix.discovery.TimedSupervisorTask  : task supervisor timed out |  
  |   |  
  | java.util.concurrent.TimeoutException: null |  
  | at java.base/java.util.concurrent.FutureTask.get(FutureTask.java:204) |  
  | at com.netflix.discovery.TimedSupervisorTask.run(TimedSupervisorTask.java:65) |  
  | at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:572) |  
  | at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317) |  
  | at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304) |  
  | at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144) |  
  | at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642) |  
  | at java.base/java.lang.Thread.run(Thread.java:1583) |  
  |   |  
  | 2025-08-07T18:40:20.590Z  WARN 1 --- [Artemis] [coveryClient-%d] c.netflix.discovery.TimedSupervisorTask  : task supervisor timed out |  
  |   |  
  | java.util.concurrent.TimeoutException: null |  
  | at java.base/java.util.concurrent.FutureTask.get(FutureTask.java:204) |  
  | at com.netflix.discovery.TimedSupervisorTask.run(TimedSupervisorTask.java:65) |  
  | at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:572) |  
  | at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317) |  
  | at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304) |  
  | at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144) |  
  | at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642) |  
  | at java.base/java.lang.Thread.run(Thread.java:1583) |  
  |   |  
  | 2025-08-07T18:40:25.745Z  INFO 1 --- [Artemis] [at-handler-2920] t.c.a.a.p.s.ProgrammingSubmissionService : processNewProgrammingSubmission invoked due to the commit 6e8bc9a4c1478231ff54cefdd8df57086e76289b by Florian Glombik with florian.glombik@tum.de in branch main |  
  | 2025-08-07T18:40:25.745Z  INFO 1 --- [Artemis] [at-handler-2920] d.t.c.a.a.p.s.l.LocalCITriggerService    : Triggering build for participation 14656 and commit hash 6e8bc9a4c1478231ff54cefdd8df57086e76289b |  
  | 2025-08-07T18:40:25.755Z ERROR 1 --- [Artemis] [at-handler-2920] d.t.c.a.a.p.service.AbstractGitService   : Cannot use ssh properly because of mismatch of Jgit transport object: org.eclipse.jgit.transport.TransportLocal@7b3ddc65 |  
  | 2025-08-07T18:40:25.885Z ERROR 1 --- [Artemis] [at-handler-2920] d.t.c.a.a.p.service.AbstractGitService   : Cannot use ssh properly because of mismatch of Jgit transport object: org.eclipse.jgit.transport.TransportLocal@276b87f |  
  | 2025-08-07T18:40:26.491Z  INFO 1 --- [Artemis] [at-handler-2920] d.t.c.a.a.p.s.l.LocalCITriggerService    : Added build job 146561754592026002 for exercise test and participation 14656 with priority 2 to the queue |  
  | 2025-08-07T18:40:26.662Z  INFO 1 --- [Artemis] [at-handler-2920] t.c.a.a.p.s.ProgrammingSubmissionService : Create new programmingSubmission with commitHash: 6e8bc9a4c1478231ff54cefdd8df57086e76289b for participation: 14656 |  
  | 2025-08-07T18:40:39.523Z  INFO 1 --- [Artemis] [at-handler-3002] d.t.c.a.a.p.service.GitService           : Pull /opt/artemis/data/repos/GE25BASTESTCOURSETEST/ge25bastestcoursetest-ge25bas |  
  | 2025-08-07T18:40:39.555Z ERROR 1 --- [Artemis] [at-handler-3002] d.t.c.a.a.p.service.AbstractGitService   : Cannot use ssh properly because of mismatch of Jgit transport object: org.eclipse.jgit.transport.TransportLocal@396d298b |  
  | 2025-08-07T18:40:44.392Z ERROR 1 --- [Artemis] [at-handler-3002] d.t.c.a.a.p.service.AbstractGitService   : Cannot use ssh properly because of mismatch of Jgit transport object: org.eclipse.jgit.transport.TransportLocal@46b92258 |  
  | 2025-08-07T18:40:45.348Z  INFO 1 --- [Artemis] [at-handler-3002] t.c.a.a.p.s.ProgrammingSubmissionService : processNewProgrammingSubmission invoked due to the commit a464460071a5abaa8ae49a4f604427fe1711cd92 by Florian Glombik with florian.glombik@tum.de in branch main |  
  | 2025-08-07T18:40:45.353Z  INFO 1 --- [Artemis] [at-handler-3002] d.t.c.a.a.p.s.l.LocalCITriggerService    : Triggering build for participation 14656 and commit hash a464460071a5abaa8ae49a4f604427fe1711cd92 |  
  | 2025-08-07T18:40:45.372Z ERROR 1 --- [Artemis] [at-handler-3002] d.t.c.a.a.p.service.AbstractGitService   : Cannot use ssh properly because of mismatch of Jgit transport object: org.eclipse.jgit.transport.TransportLocal@72f7b327 |  
  | 2025-08-07T18:40:45.425Z ERROR 1 --- [Artemis] [at-handler-3002] d.t.c.a.a.p.service.AbstractGitService   : Cannot use ssh properly because of mismatch of Jgit transport object: org.eclipse.jgit.transport.TransportLocal@409614e0 |  
  | 2025-08-07T18:40:45.554Z  INFO 1 --- [Artemis] [at-handler-3002] d.t.c.a.a.p.s.l.LocalCITriggerService    : Added build job 146561754592045482 for exercise test and participation 14656 with priority 2 to the queue |  
  | 2025-08-07T18:40:45.583Z  INFO 1 --- [Artemis] [at-handler-3002] t.c.a.a.p.s.ProgrammingSubmissionService : Create new programmingSubmission with commitHash: a464460071a5abaa8ae49a4f604427fe1711cd92 for participation: 14656 |  
  | 2025-08-07T18:40:48.422Z  INFO 1 --- [Artemis] [at-handler-3010] d.t.c.a.a.p.service.GitService           : Pull /opt/artemis/data/repos/GE25BASTESTCOURSETEST/ge25bastestcoursetest-ge25bas |  
  | 2025-08-07T18:40:48.471Z ERROR 1 --- [Artemis] [at-handler-3010] d.t.c.a.a.p.service.AbstractGitService   : Cannot use ssh properly because of mismatch of Jgit transport object: org.eclipse.jgit.transport.TransportLocal@d1fe133 |  
  | 2025-08-07T18:40:49.012Z ERROR 1 --- [Artemis] [at-handler-3010] d.t.c.a.a.p.service.AbstractGitService   : Cannot use ssh properly because of mismatch of Jgit transport object: org.eclipse.jgit.transport.TransportLocal@4d9e65fa |  
  | 2025-08-07T18:40:49.927Z  INFO 1 --- [Artemis] [at-handler-3010] t.c.a.a.p.s.ProgrammingSubmissionService : processNewProgrammingSubmission invoked due to the commit 9083efc037297eb02b3f0d7f8b85bdae6adc4967 by Florian Glombik with florian.glombik@tum.de in branch main |  
  | 2025-08-07T18:40:49.927Z  INFO 1 --- [Artemis] [at-handler-3010] d.t.c.a.a.p.s.l.LocalCITriggerService    : Triggering build for participation 14656 and commit hash 9083efc037297eb02b3f0d7f8b85bdae6adc4967 |  
  | 2025-08-07T18:40:49.933Z ERROR 1 --- [Artemis] [at-handler-3010] d.t.c.a.a.p.service.AbstractGitService   : Cannot use ssh properly because of mismatch of Jgit transport object: org.eclipse.jgit.transport.TransportLocal@a72b6d8 |  
  | 2025-08-07T18:40:50.031Z ERROR 1 --- [Artemis] [at-handler-3010] d.t.c.a.a.p.service.AbstractGitService   : Cannot use ssh properly because of mismatch of Jgit transport object: org.eclipse.jgit.transport.TransportLocal@6b05b8f6 |  
  | 2025-08-07T18:40:50.206Z  INFO 1 --- [Artemis] [at-handler-3010] d.t.c.a.a.p.s.l.LocalCITriggerService    : Added build job 146561754592050109 for exercise test and participation 14656 with priority 2 to the queue |  
  | 2025-08-07T18:40:50.232Z  INFO 1 --- [Artemis] [at-handler-3010] t.c.a.a.p.s.ProgrammingSubmissionService : Create new programmingSubmission with commitHash: 9083efc037297eb02b3f0d7f8b85bdae6adc4967 for participation: 14656 |  
  | 2025-08-07T18:40:50.480Z  INFO 1 --- [Artemis] [Artemis.event-3] t.c.a.a.b.s.SharedQueueProcessingService : Processing build job: BuildJobQueueItem[id=146561754592050109, name=GE25BASTESTCOURSETEST-GE25BAS, buildAgent=BuildAgentDTO[name=artemis-node-1, memberAddress=[artemis-app-node-1]:5701, displayName=Artemis Node 1], participationId=14656, courseId=19, exerciseId=46, retryCount=0, priority=2, status=null, repositoryInfo=RepositoryInfo[repositoryName=ge25bastestcoursetest-ge25bas, repositoryType=user, triggeredByPushTo=user, assignmentRepositoryUri=https://artemis-test3.artemis.cit.tum.de/git/GE25BASTESTCOURSETEST/ge25bastestcoursetest-ge25bas.git, testRepositoryUri=https://artemis-test3.artemis.cit.tum.de/git/GE25BASTESTCOURSETEST/ge25bastestcoursetest-tests.git, solutionRepositoryUri=null, auxiliaryRepositoryUris=[Ljava.lang.String;@3355022, auxiliaryRepositoryCheckoutDirectories=[Ljava.lang.String;@12b455a9], jobTimingInfo=JobTimingInfo[submissionDate=2025-08-07T18:40:50.109099567Z[Etc/UTC], buildStartDate=2025-08-07T18:40:50.243877529Z[Etc/UTC], buildCompletionDate=null, estimatedCompletionDate=2025-08-07T18:41:26.243838128Z[Etc/UTC], estimatedDuration=36], buildConfig=BuildConfig[buildScript=#!/bin/bash |  
  | cd /var/tmp/testing-dir |  
  | #!/usr/bin/env bash |  
  | set -e |  
  |   |  
  | gradle () { |  
  | echo '⚙️ executing gradle' |  
  | chmod +x ./gradlew |  
  | ./gradlew clean test |  
  | } |  
  |   |  
  | main () { |  
  | gradle |  
  | } |  
  |   |  
  | main "${@}" |  
  | , dockerImage=ls1tum/artemis-maven-template:java17-22, commitHashToBuild=9083efc037297eb02b3f0d7f8b85bdae6adc4967, assignmentCommitHash=9083efc037297eb02b3f0d7f8b85bdae6adc4967, testCommitHash=a68f370d1c6e25bd446672b8bb71f7be3300eb29, branch=main, programmingLanguage=JAVA, projectType=PLAIN_GRADLE, scaEnabled=false, sequentialTestRunsEnabled=false, resultPaths=[/var/tmp/testing-dir/**/test-results/test/*.xml], timeoutSeconds=0, assignmentCheckoutPath=null, testCheckoutPath=null, solutionCheckoutPath=null, dockerRunConfig=null], submissionResult=null] |  
  | 2025-08-07T18:40:50.832Z  INFO 1 --- [Artemis] [ocal-ci-build-2] d.t.c.a.a.b.s.BuildAgentDockerService    : ~~~~~~~~~~~~~~~~~~~~ Inspecting docker image ls1tum/artemis-maven-template:java17-22 ~~~~~~~~~~~~~~~~~~~~ |  
  | 2025-08-07T18:40:52.305Z  INFO 1 --- [Artemis] [at-handler-3016] d.t.c.a.a.p.service.GitService           : Pull /opt/artemis/data/repos/GE25BASTESTCOURSETEST/ge25bastestcoursetest-ge25bas |  
  | 2025-08-07T18:40:52.370Z ERROR 1 --- [Artemis] [at-handler-3016] d.t.c.a.a.p.service.AbstractGitService   : Cannot use ssh properly because of mismatch of Jgit transport object: org.eclipse.jgit.transport.TransportLocal@70b8c406 |  
  | 2025-08-07T18:40:53.294Z ERROR 1 --- [Artemis] [at-handler-3016] d.t.c.a.a.p.service.AbstractGitService   : Cannot use ssh properly because of mismatch of Jgit transport object: org.eclipse.jgit.transport.TransportLocal@576ff8d8 |  
  | 2025-08-07T18:40:54.127Z  INFO 1 --- [Artemis] [at-handler-3016] t.c.a.a.p.s.ProgrammingSubmissionService : processNewProgrammingSubmission invoked due to the commit 923c3c1c348ef38b3c77784e60e0165fc253b807 by Florian Glombik with florian.glombik@tum.de in branch main |  
  | 2025-08-07T18:40:54.127Z  INFO 1 --- [Artemis] [at-handler-3016] d.t.c.a.a.p.s.l.LocalCITriggerService    : Triggering build for participation 14656 and commit hash 923c3c1c348ef38b3c77784e60e0165fc253b807 |  
  | 2025-08-07T18:40:54.135Z ERROR 1 --- [Artemis] [at-handler-3016] d.t.c.a.a.p.service.AbstractGitService   : Cannot use ssh properly because of mismatch of Jgit transport object: org.eclipse.jgit.transport.TransportLocal@3cf53f46 |  
  | 2025-08-07T18:40:54.292Z ERROR 1 --- [Artemis] [at-handler-3016] d.t.c.a.a.p.service.AbstractGitService   : Cannot use ssh properly because of mismatch of Jgit transport object: org.eclipse.jgit.transport.TransportLocal@49906136 |  
  | 2025-08-07T18:40:54.623Z  WARN 1 --- [Artemis] [ocal-ci-build-2] org.eclipse.jgit.lib.Repository          : close() called when useCnt is already zero for Repository[./checked-out-repos/9083efc037297eb02b3f0d7f8b85bdae6adc4967/GE25BASTESTCOURSETEST/ge25bastestcoursetest-ge25bas/.git] |  
  | 2025-08-07T18:40:54.772Z  INFO 1 --- [Artemis] [at-handler-3016] d.t.c.a.a.p.s.l.LocalCITriggerService    : Added build job 146561754592054382 for exercise test and participation 14656 with priority 2 to the queue |  
  | 2025-08-07T18:40:54.791Z  INFO 1 --- [Artemis] [at-handler-3016] t.c.a.a.p.s.ProgrammingSubmissionService : Create new programmingSubmission with commitHash: 923c3c1c348ef38b3c77784e60e0165fc253b807 for participation: 14656 |  
  | 2025-08-07T18:40:56.846Z  INFO 1 --- [Artemis] [at-handler-3024] d.t.c.a.a.p.service.GitService           : Pull /opt/artemis/data/repos/GE25BASTESTCOURSETEST/ge25bastestcoursetest-ge25bas |  
  | 2025-08-07T18:40:56.873Z ERROR 1 --- [Artemis] [at-handler-3024] d.t.c.a.a.p.service.AbstractGitService   : Cannot use ssh properly because of mismatch of Jgit transport object: org.eclipse.jgit.transport.TransportLocal@1da1b460 |  
  | 2025-08-07T18:41:07.621Z ERROR 1 --- [Artemis] [at-handler-3024] d.t.c.a.a.p.service.AbstractGitService   : Cannot use ssh properly because of mismatch of Jgit transport object: org.eclipse.jgit.transport.TransportLocal@7766962f |  
 

<br class="Apple-interchange-newline">
```

Steps to reproduce 
1. Trigger many builds jobs
2. Let a agent take 5 build jobs
3. Wait until the agent took the 5 build jobs
4. While the build jobs are running reduce the number of accepted build jobs
5. See that the jobs timeout and eventually the server even breaks (happened when I did that with build agent one after breaking build agent 2 with this description)

tobias-lippert reviewed (dismissed):
code

ekayandan commented on src/main/webapp/app/buildagent/build-agent-details/build-agent-details.component.spec.ts:
```diff
@@ -104,6 +105,7 @@ describe('BuildAgentDetailsComponent', () => {
         maxNumberOfConcurrentBuildJobs: 2,
         numberOfCurrentBuildJobs: 2,
         status: BuildAgentStatus.ACTIVE,
+        maxConcurrentBuildsAllowed: 2,
     };
```
done

ekayandan commented on src/test/java/de/tum/cit/aet/artemis/shared/base/AbstractArtemisBuildAgentTest.java:
```diff
@@ -106,6 +109,9 @@ protected void mockServices() throws IOException {
 
+        ThreadPoolExecutor testExecutor = new ThreadPoolExecutor(2, 2, 0L, TimeUnit.MILLISECONDS, new LinkedBlockingQueue<>());
+        when(buildAgentConfiguration.getBuildExecutor()).thenReturn(testExecutor);
+
```
done

ekayandan commented on src/main/java/de/tum/cit/aet/artemis/buildagent/BuildAgentConfiguration.java:
```diff
@@ -77,14 +81,17 @@ public BuildAgentConfiguration(ProgrammingLanguageConfiguration programmingLangu
+        if (concurrentBuildsMaximum <= 0) {
+            concurrentBuildsMaximum = Runtime.getRuntime().availableProcessors();
+        }
     }
```
done

ekayandan commented on src/main/webapp/app/buildagent/build-agent-summary/build-agent-summary.component.ts:
```diff
@@ -71,10 +77,18 @@ export class BuildAgentSummaryComponent implements OnInit, OnDestroy {
             .reduce((sum, agent) => sum + (agent.maxNumberOfConcurrentBuildJobs || 0), 0);
         this.currentBuilds = this.buildAgents.reduce((sum, agent) => sum + (agent.numberOfCurrentBuildJobs || 0), 0);
+
+        // Initialize individual concurrency values
+        this.buildAgents.forEach((agent) => {
+            if (agent.buildAgent?.name && !this.concurrencyMap[agent.buildAgent.name]) {
+                this.concurrencyMap[agent.buildAgent.name] = agent.maxNumberOfConcurrentBuildJobs || 1;
+            }
+        });
```
done

ekayandan commented:
> Tested on ts3
> 
> Can we please remove the success toast on every update? For me as a user it is quite annoying, I do already get the immediate feedback in the UI that the value was saved, I do not need a toast in addition to it (which also blocks further adjustments) - if I do not receive an error toast I assume the value to be saved
> 
> I also managed to crash the server
> 
> ```
> 
> 2025-08-07T18:40:03.119Z  INFO 1 --- [Artemis] [at-handler-2920] d.t.c.a.a.p.service.GitService           : Pull /opt/artemis/data/repos/GE25BASTESTCOURSETEST/ge25bastestcoursetest-ge25bas |  
> -- | --
>   | 2025-08-07T18:40:03.196Z ERROR 1 --- [Artemis] [at-handler-2920] d.t.c.a.a.p.service.AbstractGitService   : Cannot use ssh properly because of mismatch of Jgit transport object: org.eclipse.jgit.transport.TransportLocal@4674f753 |  
>   | 2025-08-07T18:40:12.973Z ERROR 1 --- [Artemis] [at-handler-2920] d.t.c.a.a.p.service.AbstractGitService   : Cannot use ssh properly because of mismatch of Jgit transport object: org.eclipse.jgit.transport.TransportLocal@688dfc86 |  
>   | 2025-08-07T18:40:20.588Z  WARN 1 --- [Artemis] [coveryClient-%d] c.netflix.discovery.TimedSupervisorTask  : task supervisor timed out |  
>   |   |  
>   | java.util.concurrent.TimeoutException: null |  
>   | at java.base/java.util.concurrent.FutureTask.get(FutureTask.java:204) |  
>   | at com.netflix.discovery.TimedSupervisorTask.run(TimedSupervisorTask.java:65) |  
>   | at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:572) |  
>   | at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317) |  
>   | at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304) |  
>   | at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144) |  
>   | at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642) |  
>   | at java.base/java.lang.Thread.run(Thread.java:1583) |  
>   |   |  
>   | 2025-08-07T18:40:20.590Z  WARN 1 --- [Artemis] [coveryClient-%d] c.netflix.discovery.TimedSupervisorTask  : task supervisor timed out |  
>   |   |  
>   | java.util.concurrent.TimeoutException: null |  
>   | at java.base/java.util.concurrent.FutureTask.get(FutureTask.java:204) |  
>   | at com.netflix.discovery.TimedSupervisorTask.run(TimedSupervisorTask.java:65) |  
>   | at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:572) |  
>   | at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317) |  
>   | at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304) |  
>   | at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144) |  
>   | at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642) |  
>   | at java.base/java.lang.Thread.run(Thread.java:1583) |  
>   |   |  
>   | 2025-08-07T18:40:25.745Z  INFO 1 --- [Artemis] [at-handler-2920] t.c.a.a.p.s.ProgrammingSubmissionService : processNewProgrammingSubmission invoked due to the commit 6e8bc9a4c1478231ff54cefdd8df57086e76289b by Florian Glombik with florian.glombik@tum.de in branch main |  
>   | 2025-08-07T18:40:25.745Z  INFO 1 --- [Artemis] [at-handler-2920] d.t.c.a.a.p.s.l.LocalCITriggerService    : Triggering build for participation 14656 and commit hash 6e8bc9a4c1478231ff54cefdd8df57086e76289b |  
>   | 2025-08-07T18:40:25.755Z ERROR 1 --- [Artemis] [at-handler-2920] d.t.c.a.a.p.service.AbstractGitService   : Cannot use ssh properly because of mismatch of Jgit transport object: org.eclipse.jgit.transport.TransportLocal@7b3ddc65 |  
>   | 2025-08-07T18:40:25.885Z ERROR 1 --- [Artemis] [at-handler-2920] d.t.c.a.a.p.service.AbstractGitService   : Cannot use ssh properly because of mismatch of Jgit transport object: org.eclipse.jgit.transport.TransportLocal@276b87f |  
>   | 2025-08-07T18:40:26.491Z  INFO 1 --- [Artemis] [at-handler-2920] d.t.c.a.a.p.s.l.LocalCITriggerService    : Added build job 146561754592026002 for exercise test and participation 14656 with priority 2 to the queue |  
>   | 2025-08-07T18:40:26.662Z  INFO 1 --- [Artemis] [at-handler-2920] t.c.a.a.p.s.ProgrammingSubmissionService : Create new programmingSubmission with commitHash: 6e8bc9a4c1478231ff54cefdd8df57086e76289b for participation: 14656 |  
>   | 2025-08-07T18:40:39.523Z  INFO 1 --- [Artemis] [at-handler-3002] d.t.c.a.a.p.service.GitService           : Pull /opt/artemis/data/repos/GE25BASTESTCOURSETEST/ge25bastestcoursetest-ge25bas |  
>   | 2025-08-07T18:40:39.555Z ERROR 1 --- [Artemis] [at-handler-3002] d.t.c.a.a.p.service.AbstractGitService   : Cannot use ssh properly because of mismatch of Jgit transport object: org.eclipse.jgit.transport.TransportLocal@396d298b |  
>   | 2025-08-07T18:40:44.392Z ERROR 1 --- [Artemis] [at-handler-3002] d.t.c.a.a.p.service.AbstractGitService   : Cannot use ssh properly because of mismatch of Jgit transport object: org.eclipse.jgit.transport.TransportLocal@46b92258 |  
>   | 2025-08-07T18:40:45.348Z  INFO 1 --- [Artemis] [at-handler-3002] t.c.a.a.p.s.ProgrammingSubmissionService : processNewProgrammingSubmission invoked due to the commit a464460071a5abaa8ae49a4f604427fe1711cd92 by Florian Glombik with florian.glombik@tum.de in branch main |  
>   | 2025-08-07T18:40:45.353Z  INFO 1 --- [Artemis] [at-handler-3002] d.t.c.a.a.p.s.l.LocalCITriggerService    : Triggering build for participation 14656 and commit hash a464460071a5abaa8ae49a4f604427fe1711cd92 |  
>   | 2025-08-07T18:40:45.372Z ERROR 1 --- [Artemis] [at-handler-3002] d.t.c.a.a.p.service.AbstractGitService   : Cannot use ssh properly because of mismatch of Jgit transport object: org.eclipse.jgit.transport.TransportLocal@72f7b327 |  
>   | 2025-08-07T18:40:45.425Z ERROR 1 --- [Artemis] [at-handler-3002] d.t.c.a.a.p.service.AbstractGitService   : Cannot use ssh properly because of mismatch of Jgit transport object: org.eclipse.jgit.transport.TransportLocal@409614e0 |  
>   | 2025-08-07T18:40:45.554Z  INFO 1 --- [Artemis] [at-handler-3002] d.t.c.a.a.p.s.l.LocalCITriggerService    : Added build job 146561754592045482 for exercise test and participation 14656 with priority 2 to the queue |  
>   | 2025-08-07T18:40:45.583Z  INFO 1 --- [Artemis] [at-handler-3002] t.c.a.a.p.s.ProgrammingSubmissionService : Create new programmingSubmission with commitHash: a464460071a5abaa8ae49a4f604427fe1711cd92 for participation: 14656 |  
>   | 2025-08-07T18:40:48.422Z  INFO 1 --- [Artemis] [at-handler-3010] d.t.c.a.a.p.service.GitService           : Pull /opt/artemis/data/repos/GE25BASTESTCOURSETEST/ge25bastestcoursetest-ge25bas |  
>   | 2025-08-07T18:40:48.471Z ERROR 1 --- [Artemis] [at-handler-3010] d.t.c.a.a.p.service.AbstractGitService   : Cannot use ssh properly because of mismatch of Jgit transport object: org.eclipse.jgit.transport.TransportLocal@d1fe133 |  
>   | 2025-08-07T18:40:49.012Z ERROR 1 --- [Artemis] [at-handler-3010] d.t.c.a.a.p.service.AbstractGitService   : Cannot use ssh properly because of mismatch of Jgit transport object: org.eclipse.jgit.transport.TransportLocal@4d9e65fa |  
>   | 2025-08-07T18:40:49.927Z  INFO 1 --- [Artemis] [at-handler-3010] t.c.a.a.p.s.ProgrammingSubmissionService : processNewProgrammingSubmission invoked due to the commit 9083efc037297eb02b3f0d7f8b85bdae6adc4967 by Florian Glombik with florian.glombik@tum.de in branch main |  
>   | 2025-08-07T18:40:49.927Z  INFO 1 --- [Artemis] [at-handler-3010] d.t.c.a.a.p.s.l.LocalCITriggerService    : Triggering build for participation 14656 and commit hash 9083efc037297eb02b3f0d7f8b85bdae6adc4967 |  
>   | 2025-08-07T18:40:49.933Z ERROR 1 --- [Artemis] [at-handler-3010] d.t.c.a.a.p.service.AbstractGitService   : Cannot use ssh properly because of mismatch of Jgit transport object: org.eclipse.jgit.transport.TransportLocal@a72b6d8 |  
>   | 2025-08-07T18:40:50.031Z ERROR 1 --- [Artemis] [at-handler-3010] d.t.c.a.a.p.service.AbstractGitService   : Cannot use ssh properly because of mismatch of Jgit transport object: org.eclipse.jgit.transport.TransportLocal@6b05b8f6 |  
>   | 2025-08-07T18:40:50.206Z  INFO 1 --- [Artemis] [at-handler-3010] d.t.c.a.a.p.s.l.LocalCITriggerService    : Added build job 146561754592050109 for exercise test and participation 14656 with priority 2 to the queue |  
>   | 2025-08-07T18:40:50.232Z  INFO 1 --- [Artemis] [at-handler-3010] t.c.a.a.p.s.ProgrammingSubmissionService : Create new programmingSubmission with commitHash: 9083efc037297eb02b3f0d7f8b85bdae6adc4967 for participation: 14656 |  
>   | 2025-08-07T18:40:50.480Z  INFO 1 --- [Artemis] [Artemis.event-3] t.c.a.a.b.s.SharedQueueProcessingService : Processing build job: BuildJobQueueItem[id=146561754592050109, name=GE25BASTESTCOURSETEST-GE25BAS, buildAgent=BuildAgentDTO[name=artemis-node-1, memberAddress=[artemis-app-node-1]:5701, displayName=Artemis Node 1], participationId=14656, courseId=19, exerciseId=46, retryCount=0, priority=2, status=null, repositoryInfo=RepositoryInfo[repositoryName=ge25bastestcoursetest-ge25bas, repositoryType=user, triggeredByPushTo=user, assignmentRepositoryUri=https://artemis-test3.artemis.cit.tum.de/git/GE25BASTESTCOURSETEST/ge25bastestcoursetest-ge25bas.git, testRepositoryUri=https://artemis-test3.artemis.cit.tum.de/git/GE25BASTESTCOURSETEST/ge25bastestcoursetest-tests.git, solutionRepositoryUri=null, auxiliaryRepositoryUris=[Ljava.lang.String;@3355022, auxiliaryRepositoryCheckoutDirectories=[Ljava.lang.String;@12b455a9], jobTimingInfo=JobTimingInfo[submissionDate=2025-08-07T18:40:50.109099567Z[Etc/UTC], buildStartDate=2025-08-07T18:40:50.243877529Z[Etc/UTC], buildCompletionDate=null, estimatedCompletionDate=2025-08-07T18:41:26.243838128Z[Etc/UTC], estimatedDuration=36], buildConfig=BuildConfig[buildScript=#!/bin/bash |  
>   | cd /var/tmp/testing-dir |  
>   | #!/usr/bin/env bash |  
>   | set -e |  
>   |   |  
>   | gradle () { |  
>   | echo '⚙️ executing gradle' |  
>   | chmod +x ./gradlew |  
>   | ./gradlew clean test |  
>   | } |  
>   |   |  
>   | main () { |  
>   | gradle |  
>   | } |  
>   |   |  
>   | main "${@}" |  
>   | , dockerImage=ls1tum/artemis-maven-template:java17-22, commitHashToBuild=9083efc037297eb02b3f0d7f8b85bdae6adc4967, assignmentCommitHash=9083efc037297eb02b3f0d7f8b85bdae6adc4967, testCommitHash=a68f370d1c6e25bd446672b8bb71f7be3300eb29, branch=main, programmingLanguage=JAVA, projectType=PLAIN_GRADLE, scaEnabled=false, sequentialTestRunsEnabled=false, resultPaths=[/var/tmp/testing-dir/**/test-results/test/*.xml], timeoutSeconds=0, assignmentCheckoutPath=null, testCheckoutPath=null, solutionCheckoutPath=null, dockerRunConfig=null], submissionResult=null] |  
>   | 2025-08-07T18:40:50.832Z  INFO 1 --- [Artemis] [ocal-ci-build-2] d.t.c.a.a.b.s.BuildAgentDockerService    : ~~~~~~~~~~~~~~~~~~~~ Inspecting docker image ls1tum/artemis-maven-template:java17-22 ~~~~~~~~~~~~~~~~~~~~ |  
>   | 2025-08-07T18:40:52.305Z  INFO 1 --- [Artemis] [at-handler-3016] d.t.c.a.a.p.service.GitService           : Pull /opt/artemis/data/repos/GE25BASTESTCOURSETEST/ge25bastestcoursetest-ge25bas |  
>   | 2025-08-07T18:40:52.370Z ERROR 1 --- [Artemis] [at-handler-3016] d.t.c.a.a.p.service.AbstractGitService   : Cannot use ssh properly because of mismatch of Jgit transport object: org.eclipse.jgit.transport.TransportLocal@70b8c406 |  
>   | 2025-08-07T18:40:53.294Z ERROR 1 --- [Artemis] [at-handler-3016] d.t.c.a.a.p.service.AbstractGitService   : Cannot use ssh properly because of mismatch of Jgit transport object: org.eclipse.jgit.transport.TransportLocal@576ff8d8 |  
>   | 2025-08-07T18:40:54.127Z  INFO 1 --- [Artemis] [at-handler-3016] t.c.a.a.p.s.ProgrammingSubmissionService : processNewProgrammingSubmission invoked due to the commit 923c3c1c348ef38b3c77784e60e0165fc253b807 by Florian Glombik with florian.glombik@tum.de in branch main |  
>   | 2025-08-07T18:40:54.127Z  INFO 1 --- [Artemis] [at-handler-3016] d.t.c.a.a.p.s.l.LocalCITriggerService    : Triggering build for participation 14656 and commit hash 923c3c1c348ef38b3c77784e60e0165fc253b807 |  
>   | 2025-08-07T18:40:54.135Z ERROR 1 --- [Artemis] [at-handler-3016] d.t.c.a.a.p.service.AbstractGitService   : Cannot use ssh properly because of mismatch of Jgit transport object: org.eclipse.jgit.transport.TransportLocal@3cf53f46 |  
>   | 2025-08-07T18:40:54.292Z ERROR 1 --- [Artemis] [at-handler-3016] d.t.c.a.a.p.service.AbstractGitService   : Cannot use ssh properly because of mismatch of Jgit transport object: org.eclipse.jgit.transport.TransportLocal@49906136 |  
>   | 2025-08-07T18:40:54.623Z  WARN 1 --- [Artemis] [ocal-ci-build-2] org.eclipse.jgit.lib.Repository          : close() called when useCnt is already zero for Repository[./checked-out-repos/9083efc037297eb02b3f0d7f8b85bdae6adc4967/GE25BASTESTCOURSETEST/ge25bastestcoursetest-ge25bas/.git] |  
>   | 2025-08-07T18:40:54.772Z  INFO 1 --- [Artemis] [at-handler-3016] d.t.c.a.a.p.s.l.LocalCITriggerService    : Added build job 146561754592054382 for exercise test and participation 14656 with priority 2 to the queue |  
>   | 2025-08-07T18:40:54.791Z  INFO 1 --- [Artemis] [at-handler-3016] t.c.a.a.p.s.ProgrammingSubmissionService : Create new programmingSubmission with commitHash: 923c3c1c348ef38b3c77784e60e0165fc253b807 for participation: 14656 |  
>   | 2025-08-07T18:40:56.846Z  INFO 1 --- [Artemis] [at-handler-3024] d.t.c.a.a.p.service.GitService           : Pull /opt/artemis/data/repos/GE25BASTESTCOURSETEST/ge25bastestcoursetest-ge25bas |  
>   | 2025-08-07T18:40:56.873Z ERROR 1 --- [Artemis] [at-handler-3024] d.t.c.a.a.p.service.AbstractGitService   : Cannot use ssh properly because of mismatch of Jgit transport object: org.eclipse.jgit.transport.TransportLocal@1da1b460 |  
>   | 2025-08-07T18:41:07.621Z ERROR 1 --- [Artemis] [at-handler-3024] d.t.c.a.a.p.service.AbstractGitService   : Cannot use ssh properly because of mismatch of Jgit transport object: org.eclipse.jgit.transport.TransportLocal@7766962f |  
>  
> 
> <br class="Apple-interchange-newline">
> ```
> 
> Steps to reproduce
> 
> 1. Trigger many builds jobs
> 2. Let a agent take 5 build jobs
> 3. Wait until the agent took the 5 build jobs
> 4. While the build jobs are running reduce the number of accepted build jobs
> 5. See that the jobs timeout and eventually the server even breaks (happened when I did that with build agent one after breaking build agent 2 with this description)
> 
> Also, I like that I cannot adjust the number if the build agent is paused, but then I think we can remove the buttons and also add a hover text why I cannot adjust it (as I might miss that the agent is paused)

thanks a lot for testing this so thoroughly 🙏
I’ve spent quite some time trying to reproduce the crash on the server, but I wasn’t able to. The logs you shared don’t indicate a fatal error, but I believe they are signs of performance degradation. This made me think that the build threads were starving system/critical threads (like discovery client, schedulers, etc.) and eventually leading server to crash. To mitigate this, I’ve lowered the priority of build threads so that they don’t block core system operations anymore.

I will keep testing the code under high load to verify that this change stabilizes the system and will monitor closely if the issue reappears. Please let me know if you observe again or have any idea about why this might be happening, so we can investigate further.

ekayandan commented on src/main/webapp/app/buildagent/build-agent-details/build-agent-details.component.ts:
```diff
@@ -122,6 +139,7 @@ export class BuildAgentDetailsComponent implements OnInit, OnDestroy {
             }, 1000); // 1 second
             this.load();
             this.initWebsocketSubscription();
+            this.loadConcurrencyResetWarningState();
             this.searchSubscription = this.search
```
Sounds important
